### PR TITLE
MAINT/TST: fix `Slerp` typing and better iv in `Rotation`

### PR DESCRIPTION
**Pull request by cgobat**
#### Reference issue
N/A

#### What does this implement/fix?
- Adds type annotations for the `Slerp.times`, `Slerp.rotations`, and `Slerp.timedelta` attributes.
- ~~Changes `Slerp.rotvecs` to be a property rather than an attribute.~~
- Cleans up some error messaging in `Rotation.__init__()` and `Slerp.__init__()`.

### CODE REVIEW

File: scipy/spatial/transform/_rotation.pyx
Comment (tupui) [inline scipy/spatial/transform/_rotation.pyx:2837]:
```suggestion
        return (self.rotations.inv() * rotations[1:]).as_rotvec()

```

Comment (tupui) [inline scipy/spatial/transform/_rotation.pyx:2837]: This is a change of behaviour as it would recompute at every call. Can you give the rational of this change?

File: scipy/spatial/transform/_rotation.pyx
Comment (tupui) [inline scipy/spatial/transform/_rotation.pyx:2770]: I did not check but we usually have tests checking the actually message from raises. If this does not trigger anything, that would be a good sign showing a test could and should be added.

**Review by tupui — COMMENTED** [review]
Thank you @cgobat for the PR! I have some early suggestions. Let's also wait for the CI to come back.

Did you check these locally using `python dev.py mypy`?

File: scipy/spatial/transform/_rotation.pyx
Comment (cgobat) [reply scipy/spatial/transform/_rotation.pyx]: It makes more sense to me philosophically to be a property rather than an attribute calculated during `__init__`. Similarly to how `as_matrix()`, `as_euler()`, etc. are `Rotation` methods, not `.matrix`, `.euler`, etc. attributes computed once during initialization. I will modify my statement about not changing behavior though, good point.

File: scipy/spatial/transform/_rotation.pyx
Comment (cgobat) [reply scipy/spatial/transform/_rotation.pyx:2770]: Okay, gotcha. I just combined two checks/errors that were doing the same thing for the sake of avoiding repetition.

File: scipy/spatial/transform/_rotation.pyx
Comment (tupui) [reply scipy/spatial/transform/_rotation.pyx]: I am afraid we cannot change part of a public API's behaviour if there are no imperative reasons. Our policy is to be conservative and err on the side of caution about API changes and possible backward incompatibilities. Someone could be relying on this exact behaviour for some reason, have a call to this attribute in a hot loop and the change would introduce unnecessary/unintended slow down.

But if there is another argument to make the change, then sure we can consider it 😃

File: scipy/spatial/transform/_rotation.pyx
Comment (tupui) [reply scipy/spatial/transform/_rotation.pyx:2770]: Actually if we want to really improve the input validation here, adding a check on the type of `rotations` would be valuable I think. (This would also need a test)

File: scipy/spatial/transform/tests/test_rotation.py
Comment (tupui) [inline scipy/spatial/transform/tests/test_rotation.py:1308]: This is a tight long
```suggestion
    msg = "must be a sequence of at least 2 rotations"
    with pytest.raises(ValueError, match=msg):
```

Also could you add a test for the other part of the condition? We have 2 validations, hence should have 2 independent tests.

File: scipy/spatial/transform/_rotation.pyx
Comment (tupui) [inline scipy/spatial/transform/_rotation.pyx:693]: While we are at it.
```suggestion
            raise ValueError("Expected `quat` to have shape (4,) or (N, 4), "
                             f"got {quat.shape}.")
```

Did you check if there was a test for that validation?

**Review by tupui — COMMENTED** [review]
Thank you for the update @cgobat. This is already in good shape, there are a few things we could do to make it better I think.

File: scipy/spatial/transform/_rotation.pyx
Comment (cgobat) [reply scipy/spatial/transform/_rotation.pyx:693]: I wasn't sure about code style with respect to using f-strings. I will gladly replace `.format()`s with `f`s any day of the week :)

File: scipy/spatial/transform/tests/test_rotation.py
Comment (cgobat) [reply scipy/spatial/transform/tests/test_rotation.py:1308]: Will do.

**Review by tupui — APPROVED** [review]
Thank you for the update @cgobat, this LGTM. I am letting some other part of the CI run and will merge after that.

Thank you again for the PR 🚀

**Pull request merged**
